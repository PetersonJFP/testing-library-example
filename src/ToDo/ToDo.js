import React, { useState } from "react";
import uniqid from "uniqid";

const ToDo = () => {
  const [task, setTask] = useState("");
  const [tasks, setTasks] = useState([]);

  const handleSubmit = evt => {
    evt.preventDefault();
    setTasks([...tasks, { id: uniqid(), name: task }]);
    setTask("");
  };

  const removeTask = t => () => {
    setTasks(tasks.filter(item => item.id !== t.id));
  };

  return (
    <div className="todo-container">
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          value={task}
          data-testid="new-task"
          onChange={ev => setTask(ev.target.value)}
        />
        <button type="submit" data-testid="add-task" disabled={!task.length}>
          Add Task
        </button>
      </form>
      <ul>
        <ToDoList tasks={tasks} removeTask={removeTask} />
      </ul>
    </div>
  );
};

const ToDoList = ({ tasks, removeTask }) => {
  return tasks.map(t => (
    <li key={t.id}>
      {t.name}{" "}
      <button onClick={removeTask(t)} data-testid={`remove-task-${t.name}`}>
        X
      </button>
    </li>
  ));
};

export default ToDo;
