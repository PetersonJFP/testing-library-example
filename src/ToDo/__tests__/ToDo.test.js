import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";

import ToDo from "../ToDo";

describe("ToDo", () => {
  it("renders correctly", async () => {
    // render the component
    const { asFragment, getByTestId } = render(<ToDo />);

    // snapshot of the base componente, before any use
    expect(asFragment()).toMatchSnapshot("before make changes");

    // promisses to find components on generated HTML
    const input = await waitForElement(() => getByTestId("new-task"));
    const btn = await waitForElement(() => getByTestId("add-task"));

    // change an element value and click de submit button
    fireEvent.change(input, { target: { value: "foo" } });
    fireEvent.click(btn);

    fireEvent.change(input, { target: { value: "bar" } });
    fireEvent.click(btn);

    fireEvent.change(input, { target: { value: "foobar" } });
    fireEvent.click(btn);

    // snapshot of the component after use
    expect(asFragment()).toMatchSnapshot("with 3 todos (foo, bar, foobar)");
  });

  it("should save the inputted to-do and show then on HTML", async () => {
    const { getByTestId, getByText } = render(<ToDo />);

    const input = await waitForElement(() => getByTestId("new-task"));
    const btn = await waitForElement(() => getByTestId("add-task"));

    const newTask = "description";

    fireEvent.change(input, { target: { value: newTask } });
    fireEvent.click(btn);

    // try recovery the created component
    const taskItem = await waitForElement(() => getByText(newTask));

    // recently created component must be exist
    expect(taskItem).toBeDefined();
    // ou
    expect(taskItem).toBeInTheDocument();
  });

  it("should remove task when the user clicks on the task remove button", async () => {
    const { getByTestId, queryAllByText, queryByText } = render(<ToDo />);

    const input = await waitForElement(() => getByTestId("new-task"));
    const btn = await waitForElement(() => getByTestId("add-task"));

    const newTask = "description";

    fireEvent.change(input, { target: { value: newTask } });
    fireEvent.click(btn);

    // must have 1 task created whit the task name passed
    expect(queryAllByText(newTask).length).toBe(1);

    // get remove task button
    const btnRemove = await waitForElement(() =>
      getByTestId(`remove-task-${newTask}`)
    );

    // click on the remove task button
    fireEvent.click(btnRemove);

    // the query by text must not find anything
    expect(queryByText(newTask)).toBeNull();
  });
});
