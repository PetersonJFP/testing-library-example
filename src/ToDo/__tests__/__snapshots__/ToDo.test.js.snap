// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo renders correctly: before make changes 1`] = `
<DocumentFragment>
  <div
    class="todo-container"
  >
    <form>
      <input
        data-testid="new-task"
        type="text"
        value=""
      />
      <button
        data-testid="add-task"
        disabled=""
        type="submit"
      >
        Add Task
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;

exports[`ToDo renders correctly: with 3 todos (foo, bar, foobar) 1`] = `
<DocumentFragment>
  <div
    class="todo-container"
  >
    <form>
      <input
        data-testid="new-task"
        type="text"
        value=""
      />
      <button
        data-testid="add-task"
        disabled=""
        type="submit"
      >
        Add Task
      </button>
    </form>
    <ul>
      <li>
        foo 
        <button
          data-testid="remove-task-foo"
        >
          X
        </button>
      </li>
      <li>
        bar 
        <button
          data-testid="remove-task-bar"
        >
          X
        </button>
      </li>
      <li>
        foobar 
        <button
          data-testid="remove-task-foobar"
        >
          X
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
